Backport several style changes from Gecko

Loirooriol opened the pull request:
<!-- Please describe your changes on the following line: -->


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

<!-- Either: -->
- [X] There are tests for these changes OR
- [ ] These changes do not require tests because ___

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

Loirooriol committed:
Further changes required by Servo

Loirooriol committed:
style: Simplify StyleColor representation

There's no need for CurrentColor / Numeric variants when we can
represent them with the complex form.

Differential Revision: https://phabricator.services.mozilla.com/D106690

Loirooriol committed:
style: Move extremum lengths to the individual Size / MaxSize types

This will prevent growing them when introducing fit-content(<length>).

This can _almost_ be derived, if it wasn't because of the quirky stuff.
I think the macro is probably good enough for now but let me know if you
disagree.

Differential Revision: https://phabricator.services.mozilla.com/D106713

Loirooriol committed:
style: Implement basic color-mix() functionality, behind a pref, but exposed to chrome code

This is straight-forward and builds on the color animation code. This
implements only the <percentage> syntax, not the whole <color-adjuster>
syntax, which seems fairly more complex.

Of course, this only uses sRGB because that's all the colors we support,
but it should be feasible to extend to lab() / lch() colors once we
support those.

I believe this subset of syntax is useful and worth implementing, so
people can play with it and say if it's useful.

Differential Revision: https://phabricator.services.mozilla.com/D106698

Loirooriol committed:
style: Support image-set() on the cursor property

Differential Revision: https://phabricator.services.mozilla.com/D106745

Loirooriol committed:
style: Address spec changes re. color-mix

It was clarified that the percentages are weights, and two weights are
now allowed. Missing percentages are computed as 100% - the other or
50%. Other than that, commas are required etc, which is good since
that's how I implemented it originally.

Differential Revision: https://phabricator.services.mozilla.com/D107295

Loirooriol committed:
style: Have collect_completion_keywords return `url`  and `image-set`  for CursorImage

Differential Revision: https://phabricator.services.mozilla.com/D107381

Loirooriol committed:
style: Cherry-pick various servo changes

All formatting and not-part-of-the-gecko-build changes.

Differential Revision: https://phabricator.services.mozilla.com/D106634

Loirooriol committed:
style: Support image-set in the content property

Differential Revision: https://phabricator.services.mozilla.com/D107397

Loirooriol committed:
style: Paper over a crash in non-nightly

Differential Revision: https://phabricator.services.mozilla.com/D107400

Loirooriol committed:
style: part 1 - Support parsing ruby-position: alternate

Differential Revision: https://phabricator.services.mozilla.com/D107382

Loirooriol committed:
Further changes required by Servo

Loirooriol committed:
style: Remove layout.css.ruby.position-alternate.enabled pref

Differential Revision: https://phabricator.services.mozilla.com/D114044

Loirooriol committed:
Further changes required by Servo

Loirooriol committed:
Further changes required by Servo

Loirooriol committed:
style: Make the cascade data cache generic

We're going to use it both for UA sheets and for author styles in Shadow
DOM.

Differential Revision: https://phabricator.services.mozilla.com/D107265

Loirooriol committed:
style: Share CascadeData instances across ShadowRoots

This should be both a memory and speed win for pages using a lot of
Shadow DOM.

In order to make the cache properly work we need to start keying media query
results on the actual StyleSheetContents, as that's what we share on Gecko, but
that should all be fine.

Differential Revision: https://phabricator.services.mozilla.com/D107266

Loirooriol committed:
style: Make -moz-accent-color reflect the windows accent color

And remove the windows-specific versions of this color.

Also fix the hard-coded defaults by the colors I get in the default windows
theme.

Differential Revision: https://phabricator.services.mozilla.com/D108325

Loirooriol committed:
style: Invalidate a bit more aggressively when a pseudo-element matches, in order to also invalidate the cached pseudo-styles on the parent

This should probably be fine. If it becomes a perf issue somehow we can
implement the RESTYLE_PSEUDOS hint or what not.

Differential Revision: https://phabricator.services.mozilla.com/D108338

Loirooriol committed:
style: Respect system colors in the color and background-color properties

The reason why this doesn't work is because these styles come from
datetimebox.css, which is really an author style.

We could special-case these elements, but the approach that the CSSWG resolved
on for the new forced-colors spec is to respect system colors specified by
authors, see:

https://drafts.csswg.org/css-color-adjust-1/#forced-colors-properties

So this moves us towards that, and fixes the issue nicely.

Differential Revision: https://phabricator.services.mozilla.com/D108321

Loirooriol committed:
style: Remove unused -moz-default-appearance values -moz-mac-vibrancy-light/dark, and the system colors of the same name

Starting with macOS 10.14, the generic light/dark vibrancy is deprecated, and semantic vibrancy names are preferred.
If we ever need more vibrancy, we can add new values with semantic names.

Depends on D107910

Differential Revision: https://phabricator.services.mozilla.com/D108152

Loirooriol committed:
style: Part 1 - Ignore any properties which aren't valid for a style rule when generating CSS2Properties and testing propertie

To know the valid rules for each property, we need to put this information
into the Servo prop list and add an appropriate getter to Longhand/Shorthand.

Differential Revision: https://phabricator.services.mozilla.com/D105825

Loirooriol committed:
style: Part 2 - Add page-size attribute to nsStyleStruct and property parsing

This parsing is hidden behind the pref layout.css.page-size.enabled.

It isn't ideal that we parse this as a property, but we can't treat it as a
descriptor because of compatibility issues with other browsers. There are also
outstanding spec issues related to how descriptors like page-size are cascaded,
and whether the !important specifier is valid or not.

Differential Revision: https://phabricator.services.mozilla.com/D103958

Loirooriol committed:
Further changes required by Servo

Loirooriol committed:
style: Create a NonNegativePercentage type in values:specified

Differential Revision: https://phabricator.services.mozilla.com/D109285

Loirooriol committed:
style: Convert the font-stretch descriptor to use NonNegativePercentage

Depends on D109285

Differential Revision: https://phabricator.services.mozilla.com/D109286

Loirooriol committed:
style: Add ascent-, descent- and line-gap-override descriptors to the style system

Differential Revision: https://phabricator.services.mozilla.com/D109287

Loirooriol committed:
Further changes required by Servo

Loirooriol committed:
style: Use a single system font definition rather than two

Follow the pattern we use for system colors.

Differential Revision: https://phabricator.services.mozilla.com/D108822

Loirooriol committed:
Further changes required by Servo

Loirooriol committed:
style: Add the size-adjust descriptor to the style system

Differential Revision: https://phabricator.services.mozilla.com/D110022

Loirooriol committed:
Further changes required by Servo

Loirooriol committed:
style: add an appearance type that forces a dropshadow on menus on Windows, and uses SetWindowRgn to clip the popup to the border radius

Differential Revision: https://phabricator.services.mozilla.com/D109833

Loirooriol committed:
style: Convert proton pref checks from @supports rules to media features

This means that dynamic changes will be handled correctly, we can use
StaticPrefs, etc.

Differential Revision: https://phabricator.services.mozilla.com/D110816

Loirooriol committed:
style: Reduce boilerplate needed to add system-metric media features

There's no reason we can't just query LookAndFeel and we need to use
sSystemMetrics. In the past, LookAndFeel queries were not cached, but
this is no longer the case, so perf wise should be pretty equivalent.

Note that we don't need the NS_SUCCEEDED checks because the default
value from GetInt if the platform doesn't support it is 0 anyways.

Differential Revision: https://phabricator.services.mozilla.com/D110805

Loirooriol committed:
style: Clean up NPAPI plugin fallback behavior

Similifies use of EventStates and ObjectType/FallbackType enums since most states they represented are no longer valid with the removal of NPAPI plugins.  The state machine for (unsupported) plugin elements is now much simpler but still distinguishes between HTML fallbacks, fallbacks leading to a "BROKEN" state (e.g. failing to load the image the element refers to), and fallbacks that would simply lead the element to occupy an empty region.  The last type of fallback is behind a pref "layout.use-plugin-fallback" and is disabled by default.

Simplifying the state machine allows us to clean up nsObjectLoadingContent.  We also update many of the enums which refered to plugins, which would otherwise get confusing.

Differential Revision: https://phabricator.services.mozilla.com/D107158

Loirooriol committed:
style: Remove dead prefers-color-scheme: no preference code

It was removed in bug 1643656.

Differential Revision: https://phabricator.services.mozilla.com/D111461

Loirooriol committed:
style: Enable Proton Urlbar

Differential Revision: https://phabricator.services.mozilla.com/D110998

Loirooriol committed:
style: Clean up text-justify, and make distribute a parse-time alias

Since it's simpler, as discussed in the CSSWG issue.

Differential Revision: https://phabricator.services.mozilla.com/D111346

Loirooriol committed:
style: Remove -moz-proton-infobars media query support

Differential Revision: https://phabricator.services.mozilla.com/D110963

Loirooriol committed:
style: Remove various image-orientation related prefs

These shipped in 77.

Differential Revision: https://phabricator.services.mozilla.com/D111550

Loirooriol committed:
style: Move moz-control-character-visibility out of mako, and remove support for gecko_pref_controlled_initial_value

No behavior change, just cleanup. Actually seem this technically _adds_ some code even
though it's a cleanup, but that's mostly because of the wrapping of the
derive list.  The resulting code is simpler (more in-line with our usual
things, so I think it's an improvement).

Differential Revision: https://phabricator.services.mozilla.com/D111551

Loirooriol committed:
style: [css-pseudo] Use 'white-space: pre' by default in ::marker

Differential Revision: https://phabricator.services.mozilla.com/D107359

Loirooriol committed:
style: Add attribute names to the bloom filter

Safari does this. This reduces the runtime in the example linked from
comment 0 quite a lot (40ms on a local opt build, from ~130ms on a
release nightly build).

I added a pref because there's a slight chance of performance
regressions on pages that do not use attribute selectors, as we're now
doing more unconditional work per element (adding the attributes to the
bloom filter). But the trade-off should be worth it, I think.

Differential Revision: https://phabricator.services.mozilla.com/D111689

Loirooriol committed:
style: Enable forced-colors in UA and chrome always

Differential Revision: https://phabricator.services.mozilla.com/D112069

Loirooriol committed:
Further changes required by Servo

Loirooriol committed:
Further changes required by Servo

Loirooriol committed:
style: Add a CSS error to the console when using non-featureless :host selectors

(which would never match by definition).

Differential Revision: https://phabricator.services.mozilla.com/D111610

Loirooriol committed:
style: Hide -moz-outline-radius behind a pref

This property does nothing since bug 315209 got implemented.

Every single user that I checked was doing the same math by hand, so
hooray for good defaults :-)

Differential Revision: https://phabricator.services.mozilla.com/D112253

Loirooriol committed:
style: Support the type() function in image-set

Differential Revision: https://phabricator.services.mozilla.com/D109201

Loirooriol committed:
style: -moz-toolbar-prefers-color-scheme for dark theme detection

This adds a new @media query -moz-toolbar-prefers-color-scheme which works like
prefers-color-scheme but is set based on the browser theme rather than the OS
theme. The background colour of the toolbar is used to determine the theme
dark/light preference. This will be used for in-content common.css pages and
other UI elements that include that stylesheet in the browser-chrome through
shadow DOM.

The end result is that about: pages, infobars, and modals will now "match" the
browser theme (just light/dark mode, not LWT theming support).

Differential Revision: https://phabricator.services.mozilla.com/D111486

Loirooriol committed:
style: Fix whitespace handling in clamp() parsing

We should just restore the state when seeing whitespace and an unknown
token like when we do when finding any other token a few lines below.

This is not an issue for most callers (it's only an issue for clamp())
because the other callers use either `parse_comma_separated()` (for
min/max), or `parse_nested_block()` (for parens / nested calc()).

Both of those functions restrict the input in such a way that
is_exhausted returns true (a few lines above) and thus we parse
successfully.

Differential Revision: https://phabricator.services.mozilla.com/D112681

Loirooriol committed:
style: Garbage-collect event state bits

We had about 9 gaps / unused bits. I moved the devtools ones at the end
because I think we should be able to remove them (but separate bug).

Differential Revision: https://phabricator.services.mozilla.com/D113365

Loirooriol committed:
style: Implement accent-color in nsNativeBasicTheme

This is a new addition for CSS UI Level 4:

  https://drafts.csswg.org/css-ui-4/#widget-accent

I want to provide feedback on some spec issues, and thought it was a
kinda neat thing to prototype (it also makes testing contrast and such
with random GTK themes easier).

For now enable for Nightly only.

Differential Revision: https://phabricator.services.mozilla.com/D112312

Loirooriol committed:
style: Make the -moz-toolbar-prefers-color-scheme a tri-state

This will allow detecting the system theme, which allows fixing some of
the blocked bugs.

Note that when using the system theme we will still match light or dark
appropriately, so this shouldn't change behavior just yet.

Differential Revision: https://phabricator.services.mozilla.com/D113516

Loirooriol committed:
style: Make sure to consume the flusher when we hit the cascade data cache

So that sheets are properly marked as committed. The issue is that we
have this "committed" optimization, to avoid doing work if somebody adds
and removes an stylesheet without us flushing in the meantime:

  https://searchfox.org/mozilla-central/rev/f018480dfed4fc583703a5770a6db9ab9dc0fb99/servo/components/style/stylesheet_set.rs#308-319

The "committed" bit is set when we consume the flusher (in each(..)).

However when we hit the cache, before this patch, we wouldn't consume
it, which means that we may fail to do some full rebuilds even though we
need them.

Fix it by making sure we call each() in that case.

We add a test (the test would show a red square before this patch, and a
lime square with the fix).

Differential Revision: https://phabricator.services.mozilla.com/D113206

Loirooriol committed:
style: Don't allow whitespace between media query operator delim tokens

Differential Revision: https://phabricator.services.mozilla.com/D113648

Loirooriol committed:
style: Use saturating addition for math-depth

Differential Revision: https://phabricator.services.mozilla.com/D114070

Loirooriol committed:
style: Remove unused NS_CHANGE_COLOR_IF_SAME_AS_BG

It was introduced in bug 1024930, but it is unused since.

Differential Revision: https://phabricator.services.mozilla.com/D114389

Loirooriol committed:
style: Will-change should only create stacking contexts / containing blocks / etc if the property it'd change would apply

As per https://drafts.csswg.org/css-will-change/#will-change.

> If any non-initial value of a property would cause the element to
> generate a containing block for absolutely positioned elements,
> specifying that property in will-change must cause the element to
> generate a containing block for absolutely positioned elements.

But in this case the transform property wouldn't apply to the element so
there's no reason to create a stacking-context.

Differential Revision: https://phabricator.services.mozilla.com/D114121

Loirooriol committed:
Further changes required by Servo

Loirooriol committed:
Further changes required by Servo

Loirooriol committed:
Further changes required by Servo

Loirooriol commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit b5f378519fcd74ddf801f7b28c571ba0fd109d11 with merge 860e3c39cc62d981e945aa132f3c5730d5861d02...

Loirooriol commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 7e6cf230b850b16c37babb308260ee0248eab70d with merge 266cdf46427da13c1f1ea0be1a06bc96d8a81359...

bors-servo commented:
:broken_heart: Test failed - [checks-github](https://github.com/servo/servo/actions/runs/4998022022/jobs/8954630555)

mrobinson reviewed (approved):
Great work, @Loirooriol! Could you squash these into a single commit and update all the comments to say something like:

"Update the `style` and `selectors` crates to gecko *hash* (*date*)"

Gecko uses Mercurial, so maybe just link to the last commit you included if a hash isn't appropriate. This will make it easier to continue the update process from where you left off, if someone else needs to take over.

Loirooriol committed:
Update test expectations

mrobinson commented on components/style/servo/media_queries.rs:
```diff
@@ -202,6 +203,12 @@ impl Device {
+    /// Returns true if the given MIME type is supported
+    pub fn is_supported_mime_type(&self, mime_type: &str) -> bool {
+        // TODO: being a valid MIME type doesn't imply that it's supported.
+        mime_type.parse::<Mime>().is_ok()
```
This should probably try to do something correct even if it results in code duplication. Maybe use the list from the networking crate, with a comment that it should be kept in sync:

- "image/x-icon"
- mime::IMAGE_BMP
- mime::IMAGE_GIF
- "image/webp"
- mime::IMAGE_PNG
- mime::IMAGE_JPEG

mrobinson commented:
This looks good to me. I would disable style for the crates that fail `test-tidy` and then make one small change that I noted above (actually look at the mime type of the image). Other than that, I think you can land this when you think it is ready. Thank you!

Loirooriol committed:
Avoid complaints from ./mach test-tidy

Loirooriol commented:
- Updated test expectations
 - Ignored files in `servo-tidy.toml` to avoid unnecessary divergences
 - Didn't squash so that bisections will be easier if something breaks
 - Updated `is_supported_mime_type`  (I tried a `match` but it didn't compile):
    ```rs
       match mime_type.parse::<Mime>() {
            Ok(m) => {
                // Keep this in sync with 'image_classifer' from
                // components/net/mime_classifier.rs
                m == mime::IMAGE_BMP
                    || m == mime::IMAGE_GIF
                    || m == mime::IMAGE_PNG
                    || m == mime::IMAGE_JPEG
                    || m == "image/x-icon"
                    || m == "image/webp"
            }
            _ => false,
        }
    ```

Loirooriol commented:
@bors-servo try=wpt-2020

bors-servo commented:
:hourglass: Trying commit 4648e58f8f8406402d2927931762c9e18bba6d9d with merge 4da54457da96bfb1fb86226ad8730a78a9bdf4a1...

Loirooriol commented:
@bors-servo try=wpt

wpt-2020 was fine https://github.com/servo/servo/actions/runs/5026662065
Not sure what happened with 2013 https://github.com/servo/servo/actions/runs/5026716509

bors-servo commented:
:hourglass: Trying commit dbb92b99ad359634e8b84417b8264f84bfc5d056 with merge 6fb84120a7d9d7ea7c2e218d22b5555c4c00c475...

bors-servo commented:
:sunny: Test successful - [checks-github](https://github.com/servo/servo/actions/runs/5027709911/jobs/9018536915)
State: approved= try=True

bors-servo commented:
:sunny: Test successful - [checks-github](https://github.com/servo/servo/actions/runs/5027709911/jobs/9018536915)
State: approved= try=True

Loirooriol commented:
@bors-servo r=mrobinson

bors-servo commented:
:pushpin: Commit dbb92b9 has been approved by `mrobinson`

<!-- @bors-servo r=mrobinson dbb92b99ad359634e8b84417b8264f84bfc5d056 -->

bors-servo commented:
:hourglass: Testing commit dbb92b99ad359634e8b84417b8264f84bfc5d056 with merge 2426a38a4d11615d71c34649b83436aa1f109260...

bors-servo commented:
:sunny: Test successful - [checks-github](https://github.com/servo/servo/actions/runs/5028185146/jobs/9020544307)
Approved by: mrobinson
Pushing 2426a38a4d11615d71c34649b83436aa1f109260 to master...
